Rework filter tuning, add dynamic notch filters (#3467)
PX4/PX4-user_guide: en/config_mc/filter_tuning.md

Without the notch filter you'd have to set the low pass filter cuttoff much lower (increasing the latency) in order to avoid passing this noise to the motors.

#### Static notch filters

One or two static notch filters on the gyro sensor data that are used to filter out narrow band noise, for example harmonics at the rotor blade pass frequency.

The static notch filters can be configured using:

- First notch filter: [IMU_GYRO_NF0_BW](../advanced_config/parameter_reference.md#IMU_GYRO_NF0_BW) and [IMU_GYRO_NF0_FRQ](../advanced_config/parameter_reference.md#IMU_GYRO_NF0_FRQ).
- Second notch filter: [IMU_GYRO_NF1_BW](../advanced_config/parameter_reference.md#IMU_GYRO_NF1_BW) and [IMU_GYRO_NF1_FRQ](../advanced_config/parameter_reference.md#IMU_GYRO_NF1_FRQ).

::: info
Only two notch filters are provided.
Airframes with more than two frequency noise spikes typically clean the first two spikes with the notch filters, and subsequent spikes using the low pass filter.
:::

#### Dynamic notch filters

Dynamic notch filter are based on ESC RPM feedback and/or the onboard FFT analysis to track the rotor blade pass frequency and its harmonics.

> ESC feedback requires ESCs capable of RPM feedback such as [DShot](../peripherals/esc_motors.md#dshot) with telemetry connected, or bidirectional DShot set up (currently work in progress), or [UAVCAN/DroneCAN ESCs](../dronecan/escs.md).

Reviewer: ```suggestion
ESC feedback requires ESCs capable of RPM feedback such as [DShot](../peripherals/esc_motors.md#dshot) with telemetry connected, or bidirectional DShot set up (currently [work in progress](https://github.com/PX4/PX4-Autopilot/pull/23863)), or [UAVCAN/DroneCAN ESCs](../dronecan/escs.md).
```

## Filters

The filtering pipeline for the controllers in PX4 is described below.

::: info
Sampling and filtering is always performed at the full raw sensor rate (commonly 8kHz, depending on the IMU).
:::

### On-chip DLPF for the Gyro Sensor

This is disabled on all chips where it can be disabled (if not, cutoff frequency is set to the highest level of the chip).

### Notch Filters

Setups that have a significant lower-frequency noise spike (e.g. due to harmonics at the rotor blade pass frequency) can benefit from using the notch filter to clean the signal before it is passed to the low pass filter (these harmonics have a similar detrimental impact on motors as other sources of noise).

> Without the notch filter you'd have to set the low pass filter cutoff much lower (increasing the latency) in order to avoid passing this noise to the motors.

Reviewer 2: ```suggestion
Without the notch filter you'd have to set the low pass filter cutoff much lower (increasing the phase lag) in order to avoid passing this noise to the motors.
```

#### Static Notch Filters

> One or two static notch filters on the gyro sensor data that are used to filter out narrow band noise, for example harmonics at the rotor blade pass frequency.

Reviewer 2: ```suggestion
One or two static notch filters on the gyro sensor data that are used to filter out narrow band noise, for example a bending mode of the airframe.
```

The static notch filters can be configured using:

- First notch filter: [IMU_GYRO_NF0_BW](../advanced_config/parameter_reference.md#IMU_GYRO_NF0_BW) and [IMU_GYRO_NF0_FRQ](../advanced_config/parameter_reference.md#IMU_GYRO_NF0_FRQ).
- Second notch filter: [IMU_GYRO_NF1_BW](../advanced_config/parameter_reference.md#IMU_GYRO_NF1_BW) and [IMU_GYRO_NF1_FRQ](../advanced_config/parameter_reference.md#IMU_GYRO_NF1_FRQ).

::: info
Only two notch filters are provided.
Airframes with more than two frequency noise spikes typically clean the first two spikes with the notch filters, and subsequent spikes using the low pass filter.
:::

#### Dynamic Notch Filters

> Dynamic notch filters use ESC RPM feedback and/or the onboard FFT analysis to track the rotor blade pass frequency and its harmonics.

Reviewer: The static notch filter appears to serve the same purpose - filter blade frequency and its first harmonic (vs dynamic multiple harmonics).

1. So when would you use one or the other? 
2. The example in Daniels' original PR filtered crap from props with slight damage - I presume that results in more noise and harmonics?

Author: I suppose the dynamic one is generally better but only possible when you have ESC feedback. Not sure about the FFT one? 

@dagar when would you use the FFT instead of the static notch filter?

Reviewer: Shall we merge and you chase him for this answer offline? My concern is that he will be too busy to respond, and we end up merging this in 2 years, no better or worse off.

Author: Let me ping @bresch as well, he might know and answer quicker.

Reviewer 2: The FFT is to track a moving vibration peak (e.g.; the blade-pass frequency when ESC telemetry is not available). The static notch is usually to cover a structural mode of the airframe.
I would say that you either use the ESC feedback or the FFT to move the dynamic notches, but you can use a static notch in addition to the dynamic ones.

> Dynamic notch filters use ESC RPM feedback and/or the onboard FFT analysis to track the rotor blade pass frequency and its harmonics.

Reviewer: Should it be just one or other? The parameter seems to reflect this

```suggestion
Dynamic notch filters use ESC RPM feedback or the onboard FFT analysis to track the rotor blade pass frequency and its harmonics.
```

Author: No, you can have both, the param is a bitfield.

> ESC RPM feedback requires ESCs capable of providing RPM feedback such as [DShot](../peripherals/esc_motors.md#dshot) with telemetry connected, a bidirectional DShot set up ([work in progress](https://github.com/PX4/PX4-Autopilot/pull/23863)), or [UAVCAN/DroneCAN ESCs](../dronecan/escs.md).

Reviewer: Just FYI, I add PR links for "WIP" stuff. Makes it easy to clean up later.

| Parameter                                                                                                     | Description                                                              |
| ------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------ |
| <a href="IMU_GYRO_DNF_EN"></a>[IMU_GYRO_DNF_EN](../advanced_config/parameter_reference.md#IMU_GYRO_DNF_EN)    | Enable IMU gyro dynamic notch filtering. `0`: ESK RPM, `1`: Onboard FFT. |
| <a href="IMU_GYRO_FFT_EN"></a>[IMU_GYRO_FFT_EN](../advanced_config/parameter_reference.md#IMU_GYRO_FFT_EN)    | Enable onboard FFT (required if `IMU_GYRO_DNF_EN` is set to `1`).        |
| <a href="IMU_GYRO_DNF_MIN"></a>[IMU_GYRO_DNF_MIN](../advanced_config/parameter_reference.md#IMU_GYRO_DNF_MIN) | Minimum dynamic notch frequency in Hz.                                   |
| <a href="IMU_GYRO_DNF_BW"></a>[IMU_GYRO_DNF_BW](../advanced_config/parameter_reference.md#IMU_GYRO_DNF_BW)    | Bandwidth for each notch filter in Hz.                                   |

> | <a href="IMU_GYRO_DNF_HMC"></a>[IMU_GYRO_DNF_HMC](../advanced_config/parameter_reference.md#IMU_GYRO_NF0_BW)  | Number of harmonics to filter.                                           |

Reviewer: Are there any guidelines for deciding what you should set for these? I mean how many harmonics should you worry about? Are the default values likely to be fine for most cases?

Reviewer: It is OK to say no. I'd still merge this as "much better".

Author: Don't know.
